File: tests/cli/integration/output-modes.test.ts
/**
 * Gap closure #10 (PR6 #10, F2 epic claude-code-profiles-yhb):
 *
 * Output-mode combinatorics: every cell of the matrix
 *
 *   { NO_COLOR env, --no-color flag, --quiet, --json } × { TTY, non-TTY }
 *
 * produces the documented output. The spawn harness always runs the bin
 * non-TTY (pipes), so the TTY=true cells are exercised via the
 * `FORCE_COLOR=1` env-var hint where applicable; the no-colour pathway is
 * the documented default for non-TTY anyway, so tests below validate:
 *
 *   - non-TTY default → no ANSI escapes (plain text)
 *   - --no-color flag forces no ANSI even if FORCE_COLOR is set
 *   - NO_COLOR env forces no ANSI (any value, including empty string per
 *     https://no-color.org)
 *   - --quiet silences print() and warn() but keeps error() and exit codes
 *   - --json silences print()/warn()/error()/progress; emits one JSON object
 *   - --json on a verb that has structured output is parseable
 *   - --json + error path emits a JSON error envelope (or empty stdout,
 *     depending on contract)
 *
 * The full 2×4×2 matrix is huge — we exercise representative cells per
 * dimension, asserting the property each flag controls (escapes / silence /
 * JSON-shape) without snapshotting full bytes.
 */

import { afterEach, describe, expect, it } from "vitest";

import { makeFixture, type Fixture } from "../../helpers/fixture.js";

import { ensureBuilt, runCli } from "./spawn.js";

let fx: Fixture | undefined;
afterEach(async () => {
  if (fx) await fx.cleanup();
  fx = undefined;
});

const ANSI_ESC = /\x1b\[[0-9;]*m/;

describe("gap closure #10: output-mode combinatorics (PR6 #10)", () => {
  // ──────────────────────────────────────────────────────────────────────
  // Default (non-TTY) → no ANSI
  // ──────────────────────────────────────────────────────────────────────
  it("non-TTY default: no ANSI escapes in --version output", async () => {
    await ensureBuilt();
    const r = await runCli({ args: ["--version"] });
    expect(r.exitCode).toBe(0);
    expect(r.stdout).not.toMatch(ANSI_ESC);
  });

  it("non-TTY default: no ANSI in `init` output", async () => {
    await ensureBuilt();
    fx = await makeFixture({});
    const r = await runCli({
      args: ["--cwd", fx.projectRoot, "init", "--no-seed", "--no-hook"],
    });
    expect(r.exitCode).toBe(0);
    expect(r.stdout).not.toMatch(ANSI_ESC);
  });

  // ──────────────────────────────────────────────────────────────────────
  // --no-color flag (additive with NO_COLOR env per parse.ts:127–130)
  // ──────────────────────────────────────────────────────────────────────
  it("--no-color flag: no ANSI even when FORCE_COLOR=1 is set", async () => {
    await ensureBuilt();
    fx = await makeFixture({});
    const r = await runCli({
      args: ["--cwd", fx.projectRoot, "--no-color", "init", "--no-seed", "--no-hook"],
      env: { FORCE_COLOR: "1" },
    });
    expect(r.exitCode).toBe(0);
    expect(r.stdout).not.toMatch(ANSI_ESC);
  });

  // ──────────────────────────────────────────────────────────────────────
  // NO_COLOR env (per https://no-color.org — any value disables colour,
  // even an empty string)
  // ──────────────────────────────────────────────────────────────────────
  it("NO_COLOR env (any value) disables ANSI", async () => {
    await ensureBuilt();
    fx = await makeFixture({});
    const r = await runCli({
      args: ["--cwd", fx.projectRoot, "init", "--no-seed", "--no-hook"],
      env: { NO_COLOR: "1", FORCE_COLOR: "1" },
    });
    expect(r.exitCode).toBe(0);
    expect(r.stdout).not.toMatch(ANSI_ESC);
  });

  // Windows skip: empty-string env values are dropped or transformed by
  // the Win32 process-creation API (Node falls through to CreateProcessW,
  // which collapses `""` to "unset" in the child's block). The NO_COLOR
  // contract under empty-string is therefore unobservable on Windows.
  // Posix runners exercise it.
  it.skipIf(process.platform === "win32")(
    "NO_COLOR env empty string still disables ANSI (per no-color.org)",
    async () => {
      await ensureBuilt();
      fx = await makeFixture({});
      const r = await runCli({
        args: ["--cwd", fx.projectRoot, "init", "--no-seed", "--no-hook"],
        env: { NO_COLOR: "", FORCE_COLOR: "1" },
      });
      expect(r.exitCode).toBe(0);
      expect(r.stdout).not.toMatch(ANSI_ESC);
    },
  );

  // ──────────────────────────────────────────────────────────────────────
  // --quiet silences print() + warn(), preserves error() + exit codes
  // ──────────────────────────────────────────────────────────────────────
  it("--quiet: success path produces empty stdout, exit 0", async () => {
    await ensureBuilt();
    fx = await makeFixture({});
    const r = await runCli({
      args: ["--cwd", fx.projectRoot, "--quiet", "init", "--no-seed", "--no-hook"],
    });
    expect(r.exitCode).toBe(0);
    expect(r.stdout).toBe("");
  });

  it("-q (short) silences stdout for an argless verb", async () => {
    await ensureBuilt();
    fx = await makeFixture({});
    const r = await runCli({
      args: ["--cwd", fx.projectRoot, "-q", "list"],
    });
    expect(r.exitCode).toBe(0);
    expect(r.stdout).toBe("");
  });

  it("--quiet preserves error() and exit codes (use missing profile → exit 1, stderr non-empty)", async () => {
    await ensureBuilt();
    fx = await makeFixture({});
    const r = await runCli({
      args: ["--cwd", fx.projectRoot, "--quiet", "use", "nonexistent"],
    });
    expect(r.exitCode).toBe(1);
    // Errors are NOT silenced under --quiet (the contract).
    expect(r.stderr.length).toBeGreaterThan(0);
  });

  // ──────────────────────────────────────────────────────────────────────
  // --json mode
  // ──────────────────────────────────────────────────────────────────────
  it("--json: status emits exactly one JSON object on stdout, parseable", async () => {
    await ensureBuilt();
    fx = await makeFixture({});
    const r = await runCli({
      args: ["--cwd", fx.projectRoot, "--json", "status"],
    });
    expect(r.exitCode).toBe(0);
    // Must be one valid JSON object — not multiple, not concatenated.
    const parsed = JSON.parse(r.stdout);
    expect(typeof parsed).toBe("object");
    expect(parsed).not.toBeNull();
  });

  it("--json: list emits one JSON object on stdout", async () => {
    await ensureBuilt();
    fx = await makeFixture({
      profiles: { a: { manifest: { name: "a" }, files: { "x.md": "x\n" } } },
    });
    const r = await runCli({
      args: ["--cwd", fx.projectRoot, "--json", "list"],
    });
    expect(r.exitCode).toBe(0);
    const parsed = JSON.parse(r.stdout);
    expect(parsed).toHaveProperty("profiles");
    expect(Array.isArray(parsed.profiles)).toBe(true);
  });

  it("--json: --version emits one JSON object on stdout", async () => {
    await ensureBuilt();
    const r = await runCli({ args: ["--json", "--version"] });
    expect(r.exitCode).toBe(0);
    // dispatch.ts emits a structured payload under --json (see comment).
    const parsed = JSON.parse(r.stdout);
    expect(parsed).toHaveProperty("version");
  });

  it("--json: error path keeps stdout JSON-clean (no human chatter)", async () => {
    // Under --json, error envelope (if any) goes to stderr — stdout MUST
    // remain a valid JSON object OR be empty so consumers parsing stdout
    // never see mixed content. We pin the weaker contract (empty-or-JSON)
    // because the TS impl currently emits empty stdout on the error path
    // and a stderr message, which is the de-facto target for Go.
    await ensureBuilt();
    fx = await makeFixture({});
    const r = await runCli({
      args: ["--cwd", fx.projectRoot, "--json", "use", "nonexistent"],
    });
    expect(r.exitCode).toBe(1);
    if (r.stdout !== "") {
      // If the impl chose to emit an error-envelope JSON object, it must
      // still be parseable.
      JSON.parse(r.stdout);
    }
    // Error message is on stderr.
    expect(r.stderr.length).toBeGreaterThan(0);
  });

  // ──────────────────────────────────────────────────────────────────────
  // Combinatoric edge: --json + NO_COLOR + --no-color stacking is harmless
  // ──────────────────────────────────────────────────────────────────────
  it("stacking --json + --no-color + NO_COLOR=1 still emits parseable JSON", async () => {
    await ensureBuilt();
    fx = await makeFixture({});
    const r = await runCli({
      args: ["--cwd", fx.projectRoot, "--json", "--no-color", "status"],
      env: { NO_COLOR: "1" },
    });
    expect(r.exitCode).toBe(0);
    JSON.parse(r.stdout); // throws if invalid
  });

  it("--quiet + NO_COLOR + --no-color: success path is byte-empty", async () => {
    await ensureBuilt();
    fx = await makeFixture({});
    const r = await runCli({
      args: [
        "--cwd",
        fx.projectRoot,
        "--quiet",
        "--no-color",
        "init",
        "--no-seed",
        "--no-hook",
      ],
      env: { NO_COLOR: "1" },
    });
    expect(r.exitCode).toBe(0);
    expect(r.stdout).toBe("");
  });
});

File: tests/integration/output_modes_test.go
package integration_test

import (
	"encoding/json"
	"regexp"
	"testing"

	"github.com/htxryan/claude-code-config-profiles/tests/integration/helpers"
)

// TestOutputModes_* — IV/T7 translation of TS output-modes.test.ts
// (PR6 #10, F2). Output-mode combinatorics:
//   { NO_COLOR env, --no-color flag, --quiet, --json } × { TTY, non-TTY }
// The Go bin runs non-TTY (pipes) by default — TTY=true cells are
// represented via FORCE_COLOR=1 env hint where applicable. Go's renderer
// never emits ANSI to a non-TTY; tests assert the property each flag
// controls (escapes / silence / JSON-shape) without snapshotting bytes.

var ansiEsc = regexp.MustCompile(`\x1b\[[0-9;]*m`)

// ── Default (non-TTY) → no ANSI ─────────────────────────────────────────

func TestOutputModes_NonTTYDefaultVersionNoAnsi(t *testing.T) {
	helpers.EnsureBuilt(t)
	r := mustRun(t, helpers.SpawnOptions{Args: []string{"--version"}})
	if r.ExitCode != 0 {
		t.Fatalf("--version: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	if ansiEsc.MatchString(r.Stdout) {
		t.Errorf("--version stdout contains ANSI: %q", r.Stdout)
	}
}

func TestOutputModes_NonTTYDefaultInitNoAnsi(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{})
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "init", "--no-seed", "--no-hook"},
	})
	if r.ExitCode != 0 {
		t.Fatalf("init: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	if ansiEsc.MatchString(r.Stdout) {
		t.Errorf("init stdout contains ANSI: %q", r.Stdout)
	}
}

// ── --no-color flag (additive with NO_COLOR env) ────────────────────────

func TestOutputModes_NoColorFlagBeatsForceColor(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{})
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "--no-color", "init", "--no-seed", "--no-hook"},
		Env:  map[string]string{"FORCE_COLOR": "1"},
	})
	if r.ExitCode != 0 {
		t.Fatalf("init: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	if ansiEsc.MatchString(r.Stdout) {
		t.Errorf("--no-color stdout contains ANSI: %q", r.Stdout)
	}
}

// ── NO_COLOR env (per https://no-color.org) ─────────────────────────────

func TestOutputModes_NoColorEnvDisablesAnsi(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{})
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "init", "--no-seed", "--no-hook"},
		Env:  map[string]string{"NO_COLOR": "1", "FORCE_COLOR": "1"},
	})
	if r.ExitCode != 0 {
		t.Fatalf("init: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	if ansiEsc.MatchString(r.Stdout) {
		t.Errorf("NO_COLOR=1 stdout contains ANSI: %q", r.Stdout)
	}
}

func TestOutputModes_NoColorEmptyStringDisablesAnsi(t *testing.T) {
	if isWindows() {
		t.Skip("empty-string env values dropped by Win32 process-creation API")
	}
	helpers.EnsureBuilt(t)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{})
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "init", "--no-seed", "--no-hook"},
		Env:  map[string]string{"NO_COLOR": "", "FORCE_COLOR": "1"},
	})
	if r.ExitCode != 0 {
		t.Fatalf("init: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	if ansiEsc.MatchString(r.Stdout) {
		t.Errorf("NO_COLOR='' stdout contains ANSI: %q", r.Stdout)
	}
}

// ── --quiet silences print()/warn(); preserves error()+exit codes ───────

func TestOutputModes_QuietSuccessEmptyStdout(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{})
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "--quiet", "init", "--no-seed", "--no-hook"},
	})
	if r.ExitCode != 0 {
		t.Fatalf("--quiet init: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	if r.Stdout != "" {
		t.Errorf("--quiet stdout not empty: %q", r.Stdout)
	}
}

func TestOutputModes_ShortQuietList(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{})
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "-q", "list"},
	})
	if r.ExitCode != 0 {
		t.Fatalf("-q list: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	if r.Stdout != "" {
		t.Errorf("-q list stdout not empty: %q", r.Stdout)
	}
}

func TestOutputModes_QuietPreservesErrors(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{})
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "--quiet", "use", "nonexistent"},
	})
	if r.ExitCode != 1 {
		t.Fatalf("--quiet use missing: want 1, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	if len(r.Stderr) == 0 {
		t.Errorf("--quiet error path: stderr empty (errors must NOT be silenced)")
	}
}

// ── --json mode ─────────────────────────────────────────────────────────

func TestOutputModes_JsonStatusOneObject(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{})
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "--json", "status"},
	})
	if r.ExitCode != 0 {
		t.Fatalf("--json status: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	var parsed map[string]any
	if err := json.Unmarshal([]byte(r.Stdout), &parsed); err != nil {
		t.Fatalf("parse JSON: %v\nstdout=%q", err, r.Stdout)
	}
	if parsed == nil {
		t.Errorf("parsed is nil: %q", r.Stdout)
	}
}

func TestOutputModes_JsonListOneObject(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{
		Profiles: map[string]helpers.ProfileSpec{
			"a": {Manifest: map[string]any{"name": "a"}, Files: map[string]string{"x.md": "x\n"}},
		},
	})
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "--json", "list"},
	})
	if r.ExitCode != 0 {
		t.Fatalf("--json list: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	var parsed struct {
		Profiles []any `json:"profiles"`
	}
	if err := json.Unmarshal([]byte(r.Stdout), &parsed); err != nil {
		t.Fatalf("parse JSON: %v\nstdout=%q", err, r.Stdout)
	}
	if parsed.Profiles == nil {
		t.Errorf("missing 'profiles' array: %q", r.Stdout)
	}
}

func TestOutputModes_JsonVersion(t *testing.T) {
	helpers.EnsureBuilt(t)
	r := mustRun(t, helpers.SpawnOptions{Args: []string{"--json", "--version"}})
	if r.ExitCode != 0 {
		t.Fatalf("--json --version: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	var parsed struct {
		Version string `json:"version"`
	}
	if err := json.Unmarshal([]byte(r.Stdout), &parsed); err != nil {
		t.Fatalf("parse JSON: %v\nstdout=%q", err, r.Stdout)
	}
	if parsed.Version == "" {
		t.Errorf("--json --version: missing 'version' field: %q", r.Stdout)
	}
}

func TestOutputModes_JsonErrorPathStdoutClean(t *testing.T) {
	// Under --json, error envelope (if any) goes to stderr. Stdout must
	// stay valid-JSON OR empty so consumers parsing stdout never see
	// mixed content. Go bin emits empty stdout on the error path.
	helpers.EnsureBuilt(t)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{})
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "--json", "use", "nonexistent"},
	})
	if r.ExitCode != 1 {
		t.Fatalf("--json use missing: want 1, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	if r.Stdout != "" {
		// If the impl chose to emit an error-envelope, it must parse.
		var parsed any
		if err := json.Unmarshal([]byte(r.Stdout), &parsed); err != nil {
			t.Errorf("--json error path stdout not parseable: %v\nstdout=%q", err, r.Stdout)
		}
	}
	if len(r.Stderr) == 0 {
		t.Errorf("--json error path: stderr empty")
	}
}

// ── Combinatoric edge: stacking is harmless ─────────────────────────────

func TestOutputModes_StackingJsonNoColorEnvParseable(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{})
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "--json", "--no-color", "status"},
		Env:  map[string]string{"NO_COLOR": "1"},
	})
	if r.ExitCode != 0 {
		t.Fatalf("stacking: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	var parsed any
	if err := json.Unmarshal([]byte(r.Stdout), &parsed); err != nil {
		t.Fatalf("parse JSON: %v\nstdout=%q", err, r.Stdout)
	}
}

func TestOutputModes_StackingQuietNoColorEnvByteEmpty(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{})
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "--quiet", "--no-color", "init", "--no-seed", "--no-hook"},
		Env:  map[string]string{"NO_COLOR": "1"},
	})
	if r.ExitCode != 0 {
		t.Fatalf("stacking quiet: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	if r.Stdout != "" {
		t.Errorf("stacking quiet: stdout not empty: %q", r.Stdout)
	}
}
